cuda/struct_of_arrays.cu

```diff
@@ -0,0 +1,360 @@
+    const float distSqr = EPS2 + xdistanceSqr + ydistanceSqr + zdistanceSqr;
+    const float distSixth = distSqr * distSqr * distSqr;
+    // TODO Use sqrtf() with correct flags
+    const float invDistCube = 1.0f / sqrt(distSixth);
+    const float sts = mass_other * invDistCube * TIMESTEP;
+    *x_run += xdistanceSqr * sts;
+    *y_run += ydistanceSqr * sts;
+    *z_run += zdistanceSqr * sts;
+}
+
+
+__global__ void
+soa_update_k(float* posx, float* posy, float* posz, float* velx, float* vely, float* velz, float* mass) {
+    // one kernel per particle
+    // 1024 threads per particle
+    // Todo choose good names for vars
+    // Note 32 is size of warp
+    int id = blockIdx.x;
+    int other = threadIdx.x;
+
+    float x_run = 0;
+    float y_run = 0;
+    float z_run = 0;
+    const float posix = posx[id];
+    const float posiy = posy[id];
+    const float posiz = posz[id];
+
+    for (std::size_t j = other; j < PROBLEMSIZE; j += 1024) {
+        soa_pp_interaction(posix, posiy, posiz, posx[j], posy[j], posz[j], mass[j], &x_run, &y_run, &z_run);
+    }
+    atomicAdd(&velx[id], x_run);
+    atomicAdd(&vely[id], y_run);
+    atomicAdd(&velz[id], z_run);    
+}
+
+
+
+__global__ void
+soa_update_k_shared(float *posx, float *posy, float *posz, float *velx, float *vely, float *velz, float *mass) {
+    // one kernel per particle
+    // 1024 threads per particle
+    // Todo choose good names for vars
+    // Note 32 is size of warp
+    int id = blockIdx.x;
+    int other = threadIdx.x;
+    __shared__ float x_values[1024];
+    __shared__ float y_values[1024];
+    __shared__ float z_values[1024];
+
+    float x_run = 0;
+    float y_run = 0;
+    float z_run = 0;
+    const float posix = posx[id];
+    const float posiy = posy[id];
+    const float posiz = posz[id];
+
+
+    for (std::size_t j = other; j < PROBLEMSIZE; j += 1024) {
+        soa_pp_interaction(posix, posiy, posiz, posx[j], posy[j], posz[j], mass[j], &x_run, &y_run, &z_run);
+    }
+    x_values[other] = x_run;
+    y_values[other] = y_run;
+    z_values[other] = z_run;
+
+    // reduce to one
+    SYNC_THREADS;
+    int border = 512;
+    while (border != 0) {
+        if (other < border) {
+            x_values[other] += x_values[other + border];
+            y_values[other] += y_values[other + border];
+            z_values[other] += z_values[other + border];
+        }
+        border /= 2;
+        SYNC_THREADS;
+    }
+    // write back to global mem
+    if (other == 0) {
+        velx[id] += x_values[0];
+        vely[id] += y_values[0];
+        velz[id] += z_values[0];
+    }
+}
+
+
+
+
+
+__global__ void
+soa_update_t(float* posx, float* posy, float* posz, float* velx, float* vely, float* velz, float* mass) {
+    // one kernel for 1024 particles
+    // 1 thread per particle
+    // Todo choose good names for vars
+    // Note 32 is size of warp
+    int id = LINEAR_ID;
+
+    float x_run = 0;
+    float y_run = 0;
+    float z_run = 0;
```

Comment:
These variables are an optimization that you have not yet done in the AoS version. Namely that you accumulate the velocity updates (the changes to particle i) in registers before you write back to memory. That should still give you a large speedup for your AoS version.